src/main/scala/io/qbeast/spark/delta/writer/BlockWriter.scala

```diff
@@ -59,90 +72,74 @@ case class BlockWriter(
   /**
    * Writes rows in corresponding files
    *
-   * @param iter iterator of rows
-   * @return the sequence of files added
+   * @param iter
+   *   iterator of rows
+   * @return
+   *   the sequence of files added
    */
-  def writeRow(iter: Iterator[InternalRow]): Iterator[(AddFile, TaskStats)] = {
-    if (!iter.hasNext) {
-      return Iterator.empty
-    }
+  def writeRow(rows: Iterator[InternalRow]): Iterator[(AddFile, TaskStats)] = {
     val revision = tableChanges.updatedRevision
-    iter
-      .foldLeft[Map[CubeId, BlockContext]](Map()) { case (blocks, row) =>
-        val cubeId = revision.createCubeId(row.getBinary(qbeastColumns.cubeColumnIndex))
-        // TODO make sure this does not compromise the structure of the index
-        // It could happen than estimated weights
-        // doesn't include all the cubes present in the final indexed dataframe
-        // we save those newly added leaves with the max weight possible
-
-        val state = tableChanges.cubeState(cubeId).getOrElse(State.FLOODED)
-        val maxWeight = tableChanges.cubeWeights(cubeId).getOrElse(Weight.MaxValue)
-        val blockCtx = blocks.getOrElse(cubeId, buildWriter(cubeId, state, maxWeight))
-
-        // The row with only the original columns
-        val cleanRow = Seq.newBuilder[Any]
-        cleanRow.sizeHint(row.numFields)
-        for (i <- 0 until row.numFields) {
-          if (!qbeastColumns.contains(i)) {
-            cleanRow += row.get(i, schemaIndex(i).dataType)
-          }
+    val contexts = mutable.Map.empty[CubeId, BlockContext]
+    rows.foreach { row =>
+      val cubeId = revision.createCubeId(row.getBinary(qbeastColumns.cubeColumnIndex))
+      // TODO make sure this does not compromise the structure of the index
```

Comment:
I'd not recommend to have TODOs in the code without a corresponding issue open, otherwise it is easy to forget that such TODOs are there in the code. If there is an issue open, then add the issue number to the comment, pls.

Reply:
Aaah - this was a pretty old TODO. This is not needed anymore since the search algorithm was modified to accept jumps between branches if it was the case. I would remove it.